docs/USAGE-GUIDE.md

```diff
@@ -1373,6 +1382,78 @@ int s2n_config_add_ticket_crypto_key(struct s2n_config *config, const uint8_t *n
 **s2n_config_add_ticket_crypto_key** adds session ticket key on the server side. It would be ideal to add new keys after every (encrypt_decrypt_key_lifetime_in_nanos/2) nanos because
 this will allow for gradual and linear transition of a key from encrypt-decrypt state to decrypt-only state.
 
+### Asynchronous private key operations related calls
+
+When s2n is used in non-blocking mode, this set of functions allows user
+to move execution of CPU-heavy private key operations out of the main
+event loop, preventing **s2n_negotiate** blocking the loop for a few
+milliseconds each time the private key operation needs to be performed.
+
+To enable asynchronous private key operations user needs to provide a
+callback function **s2n_async_pkey_fn** to
+**s2n_config_set_async_pkey_callback** call. This function will be
+executed during **s2n_negotiate** call every time operation on private
+key needs to be performed. The argument **op** represents the operation
+to perform. From the callback the user can spawn the thread to perform
+**op** through **s2n_async_pkey_op_perform** call and immediately return
+**S2N_SUCCESS** from the function without waiting for thread to complete.
+The **s2n_negotiate** will return **S2N_FAILURE** with **S2N_ERR_T_BLOCKED**
+error type and **s2n_blocked_status** **S2N_BLOCKED_ON_APPLICATION_INPUT**,
+and will keep giving the same error until the **op** is performed and
+applied to the connection through **s2n_async_pkey_op_apply** call.
+
+Note, it is not safe to call multiple functions on the same **conn** or
+**op** objects from 2 different threads at the same time. Doing so will
+produce undefined behaviour. However it is safe to have a call to
+function involvingi only **conn** at the same time with a call to
+function involving only **op**, as 2 objects are not coupled with each
```

Comment:
Grammar nit: "as **those** 2 objects are not coupled with each other"

Reply:
fixed